### ISSUE TITLE
Enhancement: Support "+json" suffixed media types per RFC 6839

### ISSUE THREAD

**Issue by bunny-therapist** [litestar-org/litestar#3088]
### Summary

Json encoded responses in Litestar must have a media type which starts with "application/json", as can be seen https://github.com/litestar-org/litestar/blob/main/litestar/response/base.py#L388
```
if media_type.startswith("application/json"):
    return encode_json(content, enc_hook)
```

I would like this to change so that it supports "application" media types with the "+json" suffix, as described in  [RFC 6839](https://datatracker.ietf.org/doc/html/rfc6839#section-3.1). For example, the lines could change to something like
```
super_type, sub_type = media_type.split("/", 1)
if super_type == "application" and (sub_type.startswith("json") or sub_type.endswith("+json")):
    return encode_json(content, enc_hook)
```
So that I can set the media type to, e.g., "application/problem+json" (as described [here](https://webconcepts.info/concepts/media-type/application/problem+json)), and then json encoding would still work.

### Basic Example

I would like to be able to do something like the below and have the json encoding of the response work. If I run the below app now, the schema look alright, but the endpoint gives a server error because the response encoding fails since the media type does not start with "application/json". (It is clear that this is the problem because if I change the media type to "application/json+problem", it works.)
```python
from typing import Optional
import pydantic
import litestar
import litestar.status_codes

class ProblemDetails(pydantic.BaseModel):
    title: str
    type: str
    status: Optional[int] = None
    detail: Optional[str] = None
    instance: Optional[str] = None


@litestar.get(
    status_code=litestar.status_codes.HTTP_418_IM_A_TEAPOT,
    media_type="application/problem+json",
)
async def problem_endpoint() -> ProblemDetails:
    return ProblemDetails(title="broken", type="bad server")

app = litestar.Litestar(route_handlers=[problem_endpoint])
```

### Drawbacks and Impact

I do not see any drawbacks. If you preserve the behavior for media types which start with "application/json", I think it should all be fine. It just extends the media types which can be encoded in responses, which is very unlikely to be problematic since I doubt many servers rely on response encoding failing.

The "+json" suffix for application media types is not uncommon. For example, the "ProblemDetails" specification is quite common. Since there is an RFC for it (RFC 6839) it is clear that this is a standard worthy to support.

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/3088">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/3088/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/litestar-org/litestar/issues/3088/pledge.svg">
</picture>
</a>

**Comment by provinzkraut** [comment]
This would be a good feature to have. It seems like you have already identified what code changes would have to be made to support this, would you want to submit a PR for this?

**Comment by bunny-therapist** [comment]
Sure. I would have to familiarize myself a bit with the setup and make sure I have not overlooked anything, but I'll give it a try.

**Comment by provinzkraut** [comment]
Closed in #3096
